### PR TITLE
[cxx-interop] Allow function templates with defaulted template type parameters to be called.

### PR DESCRIPTION
**Pull request by zoecarver**
If a defaulted template type parameter is not used in the function's
signature, don't create a corresponding generic argument for that
template type. This allows us to call function templates with defaulted
template type parameters. This is very common in the standard library
for things like enable_if which is used to disable various
functions/overloads with SFINAE.

The biggest part of this change is going forward not all function
templates will be imported as generic functions in Swift. This should
work OK but we may discover there was some logic which only looked for
generic function when dealing with function templates.

### CODE REVIEW

File: test/Interop/Cxx/templates/defaulted-template-type-parameter.swift [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import StdlibUnittest
+
+// The purpose of this test is to make sure that we correctly IRGen these
+// templates and link them. The behavior is not important here (we test that
+// elsewhere).
+var DefaultedTemplateTestSuite = TestSuite("Defaulted Template Type Parameters")
+
+DefaultedTemplateTestSuite.test("Correct ctor picked") {
+  let x1 = X(0)
+  expectEqual(x1.picked, .arg)
```
Comment (zoecarver) [inline test/Interop/Cxx/templates/defaulted-template-type-parameter.swift:15]: Unrelated: the fact that this *just works* makes me so happy :)

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
cc @plotfi (I wasn't allowed to add you as a review for some reason).

**Comment by plotfi** [comment]
@zoecarver This is good stuff. I know I saw a good number of template with defaulted parameters in some of our frameworks when I audited them manually: https://github.com/facebook/componentkit/blob/master/ComponentKit/LayoutComponents/FlexboxComponentBuilder.h#L27-L30

File: lib/AST/ClangTypeConverter.cpp [hunk 2]
```diff
@@ -865,6 +865,9 @@ ClangTypeConverter::getClangTemplateArguments(
     ArrayRef<Type> genericArgs,
     SmallVectorImpl<clang::TemplateArgument> &templateArgs) {
   assert(templateArgs.size() == 0);
+  if (genericArgs.empty()) // If we have no generic argument, this is a noop.
```
Comment (plotfi) [inline lib/AST/ClangTypeConverter.cpp:868]: If we have a comment here can we explain why it is a noop? I feel like `genericArgs.empty()` communicates that sort of on it own?

**Review by egorzhdan — COMMENTED** [review]
This is great! I've also stumbled upon a few cases of weirdly imported generic clauses when playing with the C++ stdlib.
I think this patch is going to fix that.

File: lib/ClangImporter/ImportDecl.cpp [hunk 3]
```diff
@@ -4004,15 +4014,32 @@ namespace {
       if (funcTemplate) {
         unsigned i = 0;
         for (auto param : *funcTemplate->getTemplateParameters()) {
+          auto templateTypeParam = cast<clang::TemplateTypeParmDecl>(param);
+          // If the template type parameter isn't used in the signature then we
```
Comment (plotfi) [inline lib/ClangImporter/ImportDecl.cpp:4018]: Actually this makes me wonder, what happens with template functions who don't use the templated parameters at all? Would something like `template <typename T> void f() {}` be a different patch entirely or could we just pick some primitive type to use as a specialization?

File: lib/ClangImporter/ImportDecl.cpp [hunk 4]
```diff
@@ -3996,6 +3996,16 @@ namespace {
       DeclName name = accessorInfo ? DeclName() : importedName.getDeclName();
       auto selfIdx = importedName.getSelfIndex();
 
+      auto templateParamTypeUsedInSignature =
+          [&decl](clang::TemplateTypeParmDecl *type) -> bool {
+        for (unsigned i : range(0, decl->getNumParams())) {
```
[the remark this answers is not in the document]
Comment (zoecarver) [reply lib/ClangImporter/ImportDecl.cpp:4001]: No, it wouldn't. I'll update it to also check for the return type.

File: lib/ClangImporter/ImportDecl.cpp [hunk 5]
```diff
@@ -3996,6 +3996,16 @@ namespace {
...
+          if (decl->getParamDecl(i)->getType().getTypePtr() ==
```
[the remark this answers is not in the document]
Comment (zoecarver) [reply lib/ClangImporter/ImportDecl.cpp:4002]: Good question. No, it wouldn't. But we can't import these functions anyway because we can't handle dependent types yet. I have a test for this.

[the remark this answers is not in the document]
Comment (zoecarver) [reply lib/ClangImporter/ImportDecl.cpp:4002]: I'll add a TODO comment here as well referencing the SR to implement support for dependent types.

File: lib/ClangImporter/ImportDecl.cpp (hunk 3 above)
Comment (zoecarver) [reply lib/ClangImporter/ImportDecl.cpp:4018]: I think picking a random primitive would lead to users incorrectly calling `f()` in the worst case and in the best case just wouldn't be that helpful. If there's a template type param, I hope it's there for a reason. 

To support these kinds of templates, I think we talked about importing them like this:
```
func f<T>(_ _: T.Type)
```

But that will need an evolution discussion.

File: lib/ClangImporter/ImportDecl.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (egorzhdan) [reply lib/ClangImporter/ImportDecl.cpp:4002]: Ah you're right, I didn't notice the test for it. Thanks!

File: lib/AST/ClangTypeConverter.cpp (hunk 2 above)
Comment (zoecarver) [reply lib/AST/ClangTypeConverter.cpp:868]: There is actually a comment that should go here. You're right, it's not the comment that is currently there, though. I added a different comment explaining why we *have* to bail here so the assertion directly under this doesn't fire. Let me know what you think.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

File: lib/ClangImporter/ImportDecl.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (egorzhdan) [reply lib/ClangImporter/ImportDecl.cpp:4007]: I think I found a case which we might need to handle here as well:
```cpp
template <typename T = int>
void foo(T& t) {}
```
This function is imported with an `UnsafeMutablePointer<T>` argument, so `T` is used in the function signature, but the type `T&` doesn't exactly match `T`.

[the remark this answers is not in the document]
Comment (zoecarver) [reply lib/ClangImporter/ImportDecl.cpp:4007]: Good catch. I updated this patch to strip any reference/pointer from the type. 

Can you think of any other types that we currently support that might also have this issue?

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

File: lib/ClangImporter/ImportDecl.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (egorzhdan) [reply lib/ClangImporter/ImportDecl.cpp]: Nope, I can't think of any other such types. This logic looks good now 👍

**Review by egorzhdan — APPROVED** [review]
LGTM!

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
@swift-ci please test Windows.

**Comment by zoecarver** [comment]
@swift-ci please test Windows.

**Comment by zoecarver** [comment]
@swift-ci please test Windows.
